### PR TITLE
Chain sale

### PR DESCRIPTION
**Pull request by befitsandpiper**
Sale chain object that allows a fulfiller to mint a new nft in exchange for a fungible token

### CODE REVIEW

File: chain-api/src/types/sale.ts [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+export class TokenSaleDtoValidationError extends ValidationFailedError {
+  constructor(dtoName: string, errors: string[]) {
+    super(`${dtoName} validation failed: ${errors.join(". ")}`, {
+      dtoName,
+      errors
+    });
+  }
+}
+
+export class TokenSaleFailedError extends DefaultError {
```
Comment (dzikowski) [inline chain-api/src/types/sale.ts:499]: `DefaultError` ends up with `500` http code, probably a different base - maybe validation failed also? It has `400`

File: chaincode/src/sales/createTokenSale.ts [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+      const compositeKey = TokenClass.buildTokenClassCompositeKey(tokenClassKey);
+      const tokenClass = await getObjectByKey(ctx, TokenClass, compositeKey);
+
+      if(!tokenClass.authorities.includes(newSale.owner)) {
+        chainValidationErrors.push(
+          `${newSale.owner} is not an authority on ${tokenClassKey.toStringKey()}`
+        );
+      }
+
+      // TODO: Is this the correct way to check for remaining supply?
```
Comment (dzikowski) [inline chaincode/src/sales/createTokenSale.ts:184]: tbh I'm not sure - maybe @sentientforest ?

**Review by dzikowski — APPROVED** [review]
I assume this is not prod-ready, just a hackathon result to experiment on. Definitely requires testing and some cleanup before making it prod ready. Probably we can even rename the PR to something that points out that this is experimental. Like `Token sale experimental PoC` or something. This way it will be visible in release notes.

File: chaincode/src/sales/createTokenSale.ts (hunk 2 above)
Comment (befitsandpiper) [reply chaincode/src/sales/createTokenSale.ts:184]: This has been fixed, and was backwards I believe

File: chain-api/src/types/sale.ts (hunk 1 above)
Comment (befitsandpiper) [reply chain-api/src/types/sale.ts:499]: yes, it is very experimental proof of concept. I think we do eventually want this in prod, but probably requires at least another week of work to be prod ready

**Pull request merged**